src/screencast/wlr_screencast.c

```diff
@@ -295,6 +295,215 @@ struct xdpw_wlr_output *xdpw_wlr_output_first(struct wl_list *output_list) {
+
+		if (err == -1) {
+			perror("execl");
+			logprint(WARN,"Failed to execute %s",cmd);
+			return false;
+		}
+		exit(127);
+	}
+
+	close(p1[0]);
+	close(p2[1]);
+	close(p1[1]);
+	if (read(p2[0],name,name_maxlength*sizeof(char)) == -1) {
+		return false;
+	}
+
+	name[name_maxlength -1] = '\0';
+	char *p = strchr(name, '\n');
+	if (p != NULL) {
+		*p = '\0';
+	}
+	close(p2[0]);
+	return true;
+}
+
+struct xdpw_wlr_output *wlr_output_chooser_simple(char *cmd, struct wl_list *output_list) {
+
+	logprint(DEBUG, "wlroots: output chooser simple called");
+	struct xdpw_wlr_output *output, *tmp;
+	unsigned long namelength = 0;
+	int i = 0;
+
+	wl_list_for_each_safe(output, tmp, output_list, link) {
+		namelength = (namelength > strlen(output->name)) ? namelength : strlen(output->name);
+		i++;
+	}
+	namelength++;
+
+	char name[namelength];
+	memset(name, 0, namelength);
+
+	if (exec_chooser_simple(cmd, name, namelength)) {
+		logprint(TRACE, "wlroots: output chooser %s selects output %s", cmd, name);
+		wl_list_for_each_safe(output, tmp, output_list, link) {
+			if (strcmp(output->name, name) == 0) {
+				return output;
+			}
+		}
+	}
+	return NULL;
+}
+
+static bool exec_chooser_dmenu(char *cmd, char *buffer, unsigned long buffer_maxlength, char *name, unsigned long name_maxlength) {
+	int p1[2]; //p -> c
+	int p2[2]; //c -> p
+
+	if (pipe(p1) == -1) {
+		logprint(ERROR,"Failed to open pipe");
+		return false;
+	}
+	if (pipe(p2) == -1) {
+		logprint(ERROR,"Failed to open pipe");
+		return false;
+	}
+
+	pid_t pid = fork();
+
+	if (pid < 0) {
+		perror("fork");
+		return false;
+	} else if (pid == 0) {
+		close(p1[1]);
+		close(p2[0]);
+
+		dup2(p1[0], STDIN_FILENO);
+		dup2(p2[1], STDOUT_FILENO);
+		close(p1[0]);
+		close(p2[1]);
+
+		int err;
+		err = execl("/bin/sh", "/bin/sh", "-c", cmd, (void *)NULL);
+		if (err == -1) {
+			perror("execl");
+			logprint(WARN,"Failed to execute %s",cmd);
+			return false;
+		}
+		exit(127);
+	}
+	close(p1[0]);
+	close(p2[1]);
+
+	if (write(p1[1],buffer,buffer_maxlength*sizeof(char)) == -1) {
+		return false;
+	}
+	close(p1[1]);
+	if (read(p2[0],name,name_maxlength*sizeof(char)) == -1) {
+		return false;
+	}
+
+	buffer[name_maxlength -1] = '\0';
+	char *p = strchr(name, '\n');
+	if (p != NULL) {
+		*p = '\0';
+	}
+	close(p2[0]);
+	return true;
+}
+
+struct xdpw_wlr_output *wlr_output_chooser_dmenu(char *cmd, struct wl_list *output_list) {
+
+	logprint(DEBUG, "wlroots: output chooser dmenu called");
+	struct xdpw_wlr_output *output, *tmp;
+	int N = wl_list_length(output_list);
+	char *output_name[N];
+	unsigned long maxlength = 0;
+	unsigned long namelength = 0;
+	int i = 0;
+
+	logprint(TRACE, "wlroots: create outputlist %d", N);
+
+	wl_list_for_each_safe(output, tmp, output_list, link) {
+		output_name[i] = strdup(output->name);
+		maxlength += strlen(output->name)+1;
+		namelength = (namelength > strlen(output->name)) ? namelength : strlen(output->name);
+		i++;
+	}
+	namelength++;
+
+	char buffer[maxlength];
```

Comment:
It would be easier to not build a buffer, and instead directly write the output names to the pipe.